architecture/design/ysql-gin-indexes.md

```diff
@@ -0,0 +1,107 @@
+# YSQL GIN indexes
+
+As of v2.6/v2.7, Yugabyte supports only one type of index access method for DocDB-backed relations: `lsm`.  This simply maps columns to the primary key of the base table.
+
+Generalized inverted indexes map elements inside container columns to the pk of the base table.  To support GIN, we add a new access method `ybgin` and implement the access method api.  A lot of the work can be borrowed from upstream postgres's `gin` access method.
+
+## opclasses
+
+Upstream postgres comes with the following [`gin` opclasses][opclasses]:
+
+- `tsvector_ops`: map `tsvector` to `text`
+- `array_ops`: map `anyarray` to `anyelement`
+- `jsonb_ops`: map `jsonb` to `text` (key or value in jsonb)
+- `jsonb_path_ops`: map `jsonb` to `int4` (hash of root-to-leaf path in jsonb)
+
+We can just duplicate these for the `ybgin` access method and reference the same opfamily.
+
+[opclasses]: https://www.postgresql.org/docs/current/gin-builtin-opclasses.html
+
+## amapi
+
+### build
+
+`ambuild` and `yb_ambackfill` are used for `CREATE INDEX` with and without online schema changes.  They both read from the base table and insert to the index table.  Therefore, this is a superset of `yb_aminsert`.  There's not much else to it.
+
+### insert
+
+`yb_aminsert` is used instead of `aminsert` for Yugabyte.  To support index writes, extract the scan entries (reuse code from upstream) for each item and write
+
+    [scan entry, basectid]
+
+For example, given `ybgin` index on `a int[]` and insert `pk='foo', a={1,3,5}`, write
+
+    [1, 'foo']
+    [3, 'foo']
+    [5, 'foo']
+
+to the index and
+
+    ['foo'] -> '{1,3,5}'
+
+to the base table.
+
+### select
+
+`amgettuple` is used for `ybgin`, unlike `amgetbitmap` for `gin`.  Since `gin` uses bitmap scan but `ybgin` doesn't, the implementation here is most different.
+
+Given a query, extract the scan keys and entries (reuse code from upstream).  In the simplest case, only one scan key and one scan entry is given, so the query can be like
+
+    basectid := get_from_idx(scan entry)
+    return get_from_tab(basectid)
+
+A more complicated case is one key and two entries where both are required.  We can scan using one entry and recheck the condition afterward.
+
+For the first iteration, rechecking the condition is always done, even if it may not be necessary.
+
+### delete
+
+`yb_amdelete` is used.  Upstream postgres doesn't have a delete because it relies on vacuum, but we don't have that, so we need to explicitly write tombstone records.  This is similar to insert.
+
+    [scan entry, basectid] -> tombstone
+
+### update
+
+This does a delete then insert.  This may be inefficient for `ybgin` since updating one row can mean multiple deletes and inserts when only a few would have sufficed.
```

Comment:
Making it a bit more search-engine friendly with the operation name in the first sentence. This is a good pattern to follow wherever you can do it cleanly.
```suggestion
An update operation does a delete then insert. This may be inefficient for `ybgin` since updating one row can mean many deletes and inserts when only a few would have sufficed.
```

Reply:
Thanks for explaining "why" for the suggestion.